hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java

```diff
@@ -340,7 +340,10 @@ public ContainerInfo getMatchingContainer(final long size, final String owner,
       }
     } catch (Exception e) {
       LOG.warn("Container allocation failed on pipeline={}", pipeline, e);
-      return null;
+      return new ContainerInfo.Builder()
+          .setContainerID(-1)
```

Comment:
-1 is not a valid container ID.

```
IllegalStateException: -1 < 0
	at org.apache.ratis.util.Preconditions.assertTrue(Preconditions.java:73)
	at org.apache.hadoop.hdds.scm.container.ContainerInfo$Builder.setContainerID(ContainerInfo.java:398)
	at org.apache.hadoop.hdds.scm.container.ContainerManagerImpl.getMatchingContainer(ContainerManagerImpl.java:344)
	at org.apache.hadoop.hdds.scm.pipeline.WritableRatisContainerProvider.selectContainer(WritableRatisContainerProvider.java:230)
	at org.apache.hadoop.hdds.scm.pipeline.WritableRatisContainerProvider.getContainer(WritableRatisContainerProvider.java:109)
	at org.apache.hadoop.hdds.scm.pipeline.WritableContainerFactory.getContainer(WritableContainerFactory.java:74)
```

Reply:
hm, interesting. It seems 0 is a valid value and every container ID starts from 1. So I guess we could use 0 to indicate method's caller the pipelineID of the container allocation failure.